src/vs/workbench/contrib/chat/browser/chatContentParts/chatPrepareToolInvocationPart.ts

```diff
@@ -0,0 +1,117 @@
+ */
+export class ChatPrepareToolInvocationPart extends ChatProgressContentPart implements IChatContentPart {
+	private readonly element: ChatTreeItem;
+	private readonly isHiddenByFollowingContent: boolean;
+
+	constructor(
+		private prepareToolInvocation: IChatPrepareToolInvocationPart,
+		chatContentMarkdownRenderer: IMarkdownRenderer,
+		context: IChatContentPartRenderContext,
+		@IInstantiationService instantiationService: IInstantiationService,
+		@IChatMarkdownAnchorService chatMarkdownAnchorService: IChatMarkdownAnchorService,
+		@IConfigurationService configurationService: IConfigurationService,
+		@ILanguageModelToolsService private readonly languageModelToolsService: ILanguageModelToolsService,
+		@ILogService private readonly logService: ILogService
+	) {
+		// Create initial progress message - will be updated async
+		const initialProgressMessage = ChatPrepareToolInvocationPart.createInitialProgressMessage(prepareToolInvocation, languageModelToolsService);
+		super(initialProgressMessage, chatContentMarkdownRenderer, context, undefined, undefined, undefined, undefined, instantiationService, chatMarkdownAnchorService, configurationService);
+
+		this.element = context.element;
+
+		// Hide when following content contains parts that are not prepareToolInvocation or progressMessage
+		// This is similar to how progressMessage parts hide when other content arrives
+		const followingContent = context.content.slice(context.contentIndex + 1);
+		this.isHiddenByFollowingContent = followingContent.some(part => part.kind !== 'prepareToolInvocation' && part.kind !== 'progressMessage');
+
+		// Asynchronously call handleToolStream and update the message
+		if (!this.isHiddenByFollowingContent) {
+			this.fetchAndUpdateMessage(prepareToolInvocation);
+		}
+	}
+
+	private static createInitialProgressMessage(prepareToolInvocation: IChatPrepareToolInvocationPart, languageModelToolsService: ILanguageModelToolsService): IChatProgressMessage {
+		const toolData = languageModelToolsService.getTool(prepareToolInvocation.toolName);
+		const displayName = toolData?.displayName ?? prepareToolInvocation.toolName;
+		return {
+			kind: 'progressMessage',
+			content: new MarkdownString(localize('invokingTool', "Invoking tool: {0}", displayName))
+		};
+	}
+
+	private async fetchAndUpdateMessage(prepareToolInvocation: IChatPrepareToolInvocationPart): Promise<void> {
+		const toolData = this.languageModelToolsService.getTool(prepareToolInvocation.toolName);
+		if (!toolData) {
+			return;
+		}
+
+		const streamContext: IToolInvocationStreamContext = {
+			toolCallId: prepareToolInvocation.toolCallId,
+			rawInput: prepareToolInvocation.streamData?.partialInput,
+			chatRequestId: isResponseVM(this.element) ? this.element.requestId : undefined,
+			chatSessionId: this.element.sessionId
+		};
+
+		try {
+			const streamResult = await this.languageModelToolsService.handleToolStream(toolData.id, streamContext, CancellationToken.None);
```

Comment:
We cannot do side-effects like this from the chat widgets, as this will not work for background sessions.

One could say that partial results don't matter if the user isn't actively looking at the window, but we then might get into weird states if the user opens or closes the session mid-stream. Better imo to keep the view and the data model separate.